File: terraform-oci/terraform-oci/src/db/connection.ts
import oracledb, { Connection, Pool, ExecuteOptions } from 'oracledb';
import dotenv from 'dotenv';
import { resolve } from 'path';

// Load environment variables
dotenv.config();

// Database configuration interface
interface DatabaseConfig {
    user: string;
    password: string;
    connectString: string;
    walletLocation: string;
    minConnections: number;
    maxConnections: number;
    connectionTimeout: number;
    debug: boolean;
}

// Custom error class
export class DatabaseError extends Error {
    constructor(
        message: string,
        public readonly errorCode?: number,
        public readonly sqlState?: string
    ) {
        super(message);
        this.name = 'DatabaseError';
    }
}

export class DatabaseConnection {
    private static pool: Pool | null = null;
    private static config: DatabaseConfig;

    /**
     * Initialize the database configuration
     */
    private static initConfig(): DatabaseConfig {
        const config = {
            user: process.env.DB_USER || 'ADMIN',
            password: process.env.DB_PASSWORD,
            connectString: process.env.DB_DSN || 'ORCL_high',
            walletLocation: process.env.TNS_ADMIN || resolve(process.env.HOME || '', '.oracle/wallet/ORCL'),
            minConnections: parseInt(process.env.DB_MIN_CONNECTIONS || '2', 10),
            maxConnections: parseInt(process.env.DB_MAX_CONNECTIONS || '5', 10),
            connectionTimeout: parseInt(process.env.DB_CONNECTION_TIMEOUT || '60000', 10),
            debug: process.env.DB_DEBUG?.toLowerCase() === 'true'
        };

        // Validate required configurations
        if (!config.password) {
            throw new Error('Database password is required');
        }

        return config;
    }

    /**
     * Initialize the database connection pool
     */
    public static async initialize(): Promise<void> {
        if (this.pool) {
            return;
        }

        try {
            this.config = this.initConfig();

            // Configure Oracle client
            oracledb.initOracleClient({ configDir: this.config.walletLocation });
            oracledb.autoCommit = false; // Explicit transaction management

            // Create connection pool
            this.pool = await oracledb.createPool({
                user: this.config.user,
                password: this.config.password,
                connectString: this.config.connectString,
                poolMin: this.config.minConnections,
                poolMax: this.config.maxConnections,
                poolIncrement: 1,
                poolTimeout: this.config.connectionTimeout
            });

            if (this.config.debug) {
                console.log(`Connection pool created with ${this.config.minConnections} to ${this.config.maxConnections} connections`);
            }
        } catch (error) {
            throw new DatabaseError(
                `Failed to initialize database pool: ${error.message}`,
                error.errorNum,
                error.sqlState
            );
        }
    }

    /**
     * Get a connection from the pool
     */
    public static async getConnection(): Promise<Connection> {
        if (!this.pool) {
            await this.initialize();
        }
        try {
            return await this.pool!.getConnection();
        } catch (error) {
            throw new DatabaseError(
                `Failed to get database connection: ${error.message}`,
                error.errorNum,
                error.sqlState
            );
        }
    }

    /**
     * Execute a query with parameters
     */
    public static async executeQuery<T>(
        sql: string,
        params: any[] = [],
        options: ExecuteOptions = {}
    ): Promise<T[]> {
        const connection = await this.getConnection();
        try {
            const result = await connection.execute(sql, params, {
                outFormat: oracledb.OUT_FORMAT_OBJECT,
                ...options
            });
            return result.rows as T[];
        } catch (error) {
            throw new DatabaseError(
                `Query execution failed: ${error.message}`,
                error.errorNum,
                error.sqlState
            );
        } finally {
            await connection.close();
        }
    }

    /**
     * Execute a query within a transaction
     */
    public static async executeTransaction<T>(
        callback: (connection: Connection) => Promise<T>
    ): Promise<T> {
        const connection = await this.getConnection();
        try {
            const result = await callback(connection);
            await connection.commit();
            return result;
        } catch (error) {
            await connection.rollback();
            throw new DatabaseError(
                `Transaction failed: ${error.message}`,
                error.errorNum,
                error.sqlState
            );
        } finally {
            await connection.close();
        }
    }

    /**
     * Test the database connection
     */
    public static async testConnection(): Promise<boolean> {
        try {
            const result = await this.executeQuery('SELECT 1 FROM DUAL');
            return result[0]['1'] === 1;
        } catch (error) {
            console.error('Connection test failed:', error);
            return false;
        }
    }

    /**
     * Close the connection pool
     */
    public static async close(): Promise<void> {
        if (this.pool) {
            await this.pool.close(0);
            this.pool = null;
            if (this.config.debug) {
                console.log('Connection pool closed');
            }
        }
    }
}

// Example usage
async function example() {
    try {
        await DatabaseConnection.initialize();

        // Test connection
        const isConnected = await DatabaseConnection.testConnection();
        if (isConnected) {
            console.log('Successfully connected to the database!');

            // Example query
            const result = await DatabaseConnection.executeQuery('SELECT SYSDATE FROM DUAL');
            console.log('Database time:', result[0]['SYSDATE']);

            // Example transaction
            await DatabaseConnection.executeTransaction(async (connection) => {
                await connection.execute(
                    'INSERT INTO example_table (name) VALUES (:1)',
                    ['test']
                );
                return true;
            });
        }
    } catch (error) {
        console.error('Error:', error);
    } finally {
        await DatabaseConnection.close();
    }
}

if (require.main === module) {
    example().catch(console.error);
}

File: terraform-oci/terraform-oci/src/db/connection.py
import os
import cx_Oracle
from contextlib import contextmanager
from typing import Optional
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

class DatabaseConfig:
    """Database configuration class"""
    def __init__(self):
        self.user = os.getenv('DB_USER', 'ADMIN')
        self.password = os.getenv('DB_PASSWORD')
        self.dsn = os.getenv('DB_DSN', 'ORCL_high')
        self.wallet_location = os.getenv('TNS_ADMIN', os.path.expanduser('~/.oracle/wallet/ORCL'))
        self.min_connections = int(os.getenv('DB_MIN_CONNECTIONS', '2'))
        self.max_connections = int(os.getenv('DB_MAX_CONNECTIONS', '5'))
        self.connection_timeout = int(os.getenv('DB_CONNECTION_TIMEOUT', '60'))
        self.debug = os.getenv('DB_DEBUG', 'false').lower() == 'true'

        # Validate required configurations
        if not all([self.user, self.password, self.dsn, self.wallet_location]):
            raise ValueError("Missing required database configuration. Check environment variables.")

class DatabaseConnection:
    """Database connection manager"""
    _pool: Optional[cx_Oracle.SessionPool] = None
    _config: Optional[DatabaseConfig] = None

    @classmethod
    def initialize(cls, config: Optional[DatabaseConfig] = None) -> None:
        """Initialize the database connection pool"""
        if cls._pool is not None:
            return

        cls._config = config or DatabaseConfig()
        
        try:
            # Set Oracle Client configuration
            if cls._config.wallet_location:
                os.environ['TNS_ADMIN'] = cls._config.wallet_location

            # Create the connection pool
            cls._pool = cx_Oracle.SessionPool(
                user=cls._config.user,
                password=cls._config.password,
                dsn=cls._config.dsn,
                min=cls._config.min_connections,
                max=cls._config.max_connections,
                increment=1,
                encoding="UTF-8",
                timeout=cls._config.connection_timeout
            )

            if cls._config.debug:
                print(f"Connection pool created with {cls._config.min_connections} to "
                      f"{cls._config.max_connections} connections")

        except cx_Oracle.Error as error:
            raise ConnectionError(f"Failed to initialize database pool: {error}")

    @classmethod
    def get_pool(cls) -> cx_Oracle.SessionPool:
        """Get the connection pool, initializing if necessary"""
        if cls._pool is None:
            cls.initialize()
        return cls._pool

    @classmethod
    def close(cls) -> None:
        """Close the connection pool"""
        if cls._pool is not None:
            cls._pool.close()
            cls._pool = None
            if cls._config and cls._config.debug:
                print("Connection pool closed")

@contextmanager
def get_connection():
    """Context manager for database connections"""
    connection = None
    try:
        connection = DatabaseConnection.get_pool().acquire()
        yield connection
    finally:
        if connection:
            DatabaseConnection.get_pool().release(connection)

class DatabaseError(Exception):
    """Custom database error class"""
    def __init__(self, message: str, error_code: Optional[int] = None):
        self.message = message
        self.error_code = error_code
        super().__init__(self.message)

def test_connection() -> bool:
    """Test the database connection"""
    try:
        with get_connection() as connection:
            cursor = connection.cursor()
            cursor.execute("SELECT 1 FROM DUAL")
            result = cursor.fetchone()
            return result[0] == 1
    except Exception as e:
        print(f"Connection test failed: {e}")
        return False

# Example usage
if __name__ == "__main__":
    try:
        # Initialize the connection pool
        DatabaseConnection.initialize()
        
        # Test the connection
        if test_connection():
            print("Successfully connected to the database!")
            
            # Example query
            with get_connection() as connection:
                cursor = connection.cursor()
                cursor.execute("SELECT SYSDATE FROM DUAL")
                result = cursor.fetchone()
                print(f"Database time: {result[0]}")
                
    except Exception as e:
        print(f"Error: {e}")
    finally:
        DatabaseConnection.close()

